### ISSUE TITLE
https://github.com/eclipse/jetty.project/issues/4000#issue-482388946Hejsan!

### ISSUE THREAD

**Issue by firoj07** [eclipse/jetty.project#6928]
Hejsan!

The Swedish characters å ä and ö have different rules depending on where in the URL they are used.

* If they show up in the URL authority portion (host name), then ([RFC 5895](https://tools.ietf.org/html/rfc5895): IDNA 2008) punycode is preferred encoding.
* If they show up in any HTTP header (example: the `Host:` header) for HTTP, then they must be % encoded
* If they show up in the path portion of the URL then they must be % encoded
* If they show up in the query portion of the URL then they must follow form url encoding rules (which overlaps with % encoding rules)
* If they show up in the body content of a `Content-Type: application/x-www-form-urlencoded` of a HTTP form Request, then the must follow form url encoding rules (which overlaps with % encoding rules)

As an example, the UTF-8 % encoding for the 3 unicode characters ...

```
å = %C3%A5
ä = %C3%A4
ö = %C3%B6
```

The "Canonical decomposition" of [`%cc%8a` is invalid UTF-8](http://www.ltg.ed.ac.uk/~richard/utf-8.cgi?input=cc8a&mode=bytes).
Whatever charset or codepage you are using for that sequence is not compatible with today's HTTP.

Back in Jetty 6 we parsed native utf-8 and were happy with that.
In fact, back in Jetty 6 we parsed with whatever the default JVM Charset was. (which was colossally bad!  Especially so in Asian countries)
This led to mean people finding ways to inject unintended sequences of bytes to change the meaning of the request (XSS, but at the HTTP protocol level).

At around the same time the HTTP spec group at the IETF updated the HTTP/1.1 specs to make this kind of behavior not just undesirable, but outright forbidden.
From [RFC 2068 (obsolete)](https://tools.ietf.org/html/rfc2068) used back in the Jetty 6 days, to [RFC 2616 (obsolete)](https://tools.ietf.org/html/rfc2616), and again to [RFC 7230](https://tools.ietf.org/html/rfc7230).

In short, using raw UTF-8 bytes is now forbidden at the HTTP protocol level (even in modern browsers will fail many responses with raw utf-8 sequences).

_Originally posted by @joakime in https://github.com/eclipse/jetty.project/issues/4000#issuecomment-522653847_

**Comment by joakime** [comment]
Spam. Closing.
